packages/subcommands/src/lib/Subcommand.ts

```diff
@@ -0,0 +1,219 @@
+				const subcommand = mapping.subcommands.find(({ name }) => name === subcommandName);
+				if (subcommand) return this.#handleInteractionRun(interaction, context, subcommand);
+			}
+
+			if (mapping instanceof ChatInputSubcommandGroupMappings && subcommandGroupName && mapping.groupName === subcommandGroupName) {
+				const subcommand = mapping.subcommands.find(({ name }) => name === subcommandName);
+				if (subcommand) return this.#handleInteractionRun(interaction, context, subcommand);
+			}
+		}
+
+		// No match and no subcommand, return an err:
+		throw new UserError({ identifier: SubcommandPluginIdentifiers.ChatInputSubcommandNoMatch, context });
+	}
+
+	async #handleInteractionRun(
+		interaction: ChatInputCommand.Interaction,
+		context: ChatInputCommand.RunContext,
+		subcommand: ChatInputSubcommandMappingValue
+	) {
+		const payload: ChatInputSubcommandAcceptedPayload = { command: this, context, interaction };
+		const result = await fromAsync(async () => {
+			interaction.client.emit(SubcommandPluginEvents.ChatInputSubcommandRun, interaction, subcommand, payload);
+			subcommand.type ??= 'method';
+			let result: unknown;
+
+			if (subcommand.type === 'command') {
+				const parsedCommandName = subcommand.to && typeof subcommand.to === 'string' ? subcommand.to : subcommand.name;
+				const command = this.container.stores.get('commands').get(parsedCommandName);
+				if (!command?.supportsChatInputCommands())
+					throw new UserError({ identifier: SubcommandPluginIdentifiers.SubcommandNotFound, context: { ...payload } });
+
+				// Run global preconditions:
+				const globalResult = await this.container.stores
+					.get('preconditions')
+					.chatInputRun(interaction, command as ChatInputCommand, context as any);
+
+				if (!globalResult.success) {
+					this.container.client.emit(SubcommandPluginEvents.ChatInputSubcommandDenied, globalResult.error, { ...payload, subcommand });
+					return;
+				}
+
+				// Run command-specific preconditions:
+				const localResult = await command.preconditions.chatInputRun(interaction, command as ChatInputCommand, context as any);
+
+				if (!localResult.success) {
+					this.container.client.emit(SubcommandPluginEvents.ChatInputSubcommandDenied, localResult.error, { ...payload, subcommand });
+					return;
+				}
+
+				result = await command.chatInputRun(interaction, context);
+			}
+
+			if (subcommand.type === 'method' && subcommand.to) {
+				if (typeof subcommand.to === 'string') {
+					const method: ChatInputSubcommandToProperty | undefined = Reflect.get(this, subcommand.to);
+					if (!method) throw new UserError({ identifier: SubcommandPluginIdentifiers.SubcommandNotFound, context: { ...payload } });
+					result = await Reflect.apply(method, this, [interaction, context]);
+				} else {
+					result = await subcommand.to(interaction, context);
+				}
+			}
+
+			interaction.client.emit(SubcommandPluginEvents.ChatInputSubcommandSuccess, interaction, subcommand, { ...payload, result });
```

Comment:
you should always use this.container.client.emit. Also applies for message processing